### PR TITLE
Add support for deprecating packages

### PR DESCRIPTION
**Pull request by Dilhasha**
## Purpose
This PR introduces the following to support deprecating Ballerina packages.

-  Introduce bal deprecate command
- `bal pull` : print a warning when pulling a deprecated package
- `bal push` : print error when trying to push a package with a deprecated dependency.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/39203

### Valid requests

Deprecating a package:

```
> bal deprecate mynewdil/deprecate_sample:0.1.0 --message "for testing"
mynewdil/deprecate_sample:0.1.0 marked as deprecated in central successfully
```

Undo package deprecation:

```
> bal deprecate mynewdil/deprecate_sample:0.1.0 --undo
deprecation of mynewdil/deprecate_sample:0.1.0 is successfully undone in central
```

### Warnings and errors

Pull a deprecated package:

```
> bal pull mynewdil/deprecate_sample:0.1.0
WARNING [deprecate_sample] mynewdil/deprecate_sample:0.1.0 is deprecated due to : deprecating for testing
mynewdil/deprecate_sample:0.1.0 [central.ballerina.io -> home repo]  100% [====================================================================================================================================] 2/2 KB (0:00:00 / 0:00:00)
mynewdil/deprecate_sample:0.1.0 pulled from central successfully
```

`bal pack` with a dependency on a deprecated package

```
> bal pack
Compiling source
	mynewdil/deppack:1.0.0
mynewdil/deprecate_sample:0.1.0 [central.ballerina.io -> home repo]  100% [====================================================================================================================================] 2/2 KB (0:00:00 / 0:00:00)
	mynewdil/deprecate_sample:0.1.0 pulled from central successfully
WARNING [deppack] mynewdil/deprecate_sample:0.1.0 is deprecated due to : deprecating for testing

Creating bala
	target/bala/mynewdil-deppack-any-1.0.0.bala
```

Push a package with a dependency to a deprecated package

```
> bal push
mynewdil/deppack:1.0.0 [project repo -> central] 100% [===================================================================================================================================================] 2108/2108 B (0:00:00 / 0:00:00)
ballerina: unable to push package as the following dependency of the package is deprecated in the registry: mynewdil/deppack:1.0.0
```

Deprecate with unused flag:

```
> bal deprecate mynewdil/deprecate_sample:0.1.0 --undo -m "for testing"
warning: ignoring --message (-m) flag since this is an undo request
deprecation of mynewdil/deprecate_sample:0.1.0 is successfully undone in central
```

### Invalid requests
Deprecate command with invalid package(without version):

```
> bal deprecate mynewdil/deprecate_sample --message "for testing"
ballerina: invalid package. Provide the package with the version.

USAGE:
    bal deprecate {<org-name>/<package-name>:<version>} [OPTIONS]
```

Deprecate command with invalid package org:

```
> bal deprecate mynew-dil/deprecate_sample:0.1.0 --message "for testing"
ballerina: invalid organization name. Organization name can only contain alphanumeric values with lower case letters and cannot start with numeric values, org_name = mynew-dil
```

Deprecate command with invalid package name:

```
> bal deprecate mynewdil/deprecate-sample:0.1.0 --message "for testing"
ballerina: invalid package name provided. Only alphanumerics, underscores and periods are allowed in a package name and the maximum length is 256 characters, package_name = 'deprecate-sample'
```

Deprecate command with invalid package version:

```
> bal deprecate mynewdil/deprecate_sample:xxxx -m "for testing"
ballerina: invalid request received. error occurred finding package: mynewdil/deprecate_sample:xxxx_java11. Matched version is an error
```

Package not found:
```
> bal deprecate mynewdil/deprecatesample:0.1.0
ballerina: package not found for: mynewdil/deprecatesample:0.1.0
```

Trying to undo a package that is not deprecated:

```
> bal deprecate mynewdil/deprecate_sample:0.1.0 --undo
ballerina: invalid request received. package is not deprecated
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-deprecate.help [hunk 1]
```diff
@@ -0,0 +1,29 @@
+NAME
+       ballerina-deprecate - Deprecates a published package
+
+SYNOPSIS
+       bal deprecate [OPTIONS] <package>
+
+
+DESCRIPTION
+       The deprecate commands marks the specified package as deprecated in Ballerina central.
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-deprecate.help:9]:
```suggestion
       The deprecate command marks the specified package as deprecated in Ballerina central.
```

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-deprecate.help [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+       A deprecated package will not be used as a dependency of a package unless the deprecated package is already recorded in
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-deprecate.help:11]:
```suggestion
       A deprecated package will not be used as a dependency of a package unless it is already recorded in
```

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-deprecate.help [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+       A warning diagnostic will be issued whenever a deprecated package is used as a dependency.
+
+       This command does not delete the package from Ballerina central, and the package can be undeprecated using the `--undo` option.
+
+
+OPTIONS
+       --message=<msg>
+           Use the given <msg> as the deprecation message
+       -- undo
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-deprecate.help:23]:
```suggestion
       --undo
```

File: cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/DeprecateCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+ * Deprecate command tests.
+ *
+ * @since 2201.5.0
+ */
+public class DeprecateCommandTest extends BaseCommandTest {
+
+    @Test(description = "Test deprecate without package version")
+    public void testDeprecationWithoutVersion() throws IOException {
+        DeprecateCommand deprecationCommand = new DeprecateCommand(printStream, printStream, false);
+        new CommandLine(deprecationCommand).parseArgs("mynewdil/deppack", "-message", "for testing");
```
Comment (azinneera) [inline cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/DeprecateCommandTest.java:37]: Should it be `--message`?

File: cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/DeprecateCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Test(description = "Test deprecate with unused flags")
+    public void testDeprecationWithUnusedFlags() throws IOException {
+        DeprecateCommand deprecationCommand = new DeprecateCommand(printStream, printStream, false);
+        new CommandLine(deprecationCommand).parseArgs("mynewdil/deppack:0.1.0", "--undo", "-message", "for testing");
+        deprecationCommand.execute();
+
+        String buildLog = readOutput(true);
+        String actual = buildLog.replaceAll("\r", "");
+        Assert.assertTrue(actual.contains("warning: ignoring --message (-m) flag since this is an undo request"));
```
Comment (azinneera) [inline cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/DeprecateCommandTest.java:93]: Are we supporting `(-m)` shorthand?

**Review by azinneera — APPROVED** [review]

**Comment by Dilhasha** [comment]
Some error messages in Central side are fixed with https://github.com/wso2-enterprise/ballerina-registry/pull/1984

**Pull request merged**